pkg/odo/cli/init/init.go

```diff
@@ -0,0 +1,319 @@
+package init
+
+import (
+	"fmt"
+	"net/url"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/library/pkg/devfile/parser"
+	"github.com/devfile/library/pkg/devfile/parser/data/v2/common"
+
+	"github.com/redhat-developer/odo/pkg/catalog"
+	"github.com/redhat-developer/odo/pkg/devfile"
+	"github.com/redhat-developer/odo/pkg/log"
+	"github.com/redhat-developer/odo/pkg/odo/cli/init/asker"
+	"github.com/redhat-developer/odo/pkg/odo/cli/init/params"
+	"github.com/redhat-developer/odo/pkg/odo/cli/init/registry"
+	"github.com/redhat-developer/odo/pkg/odo/cmdline"
+	"github.com/redhat-developer/odo/pkg/odo/genericclioptions"
+	odoutil "github.com/redhat-developer/odo/pkg/odo/util"
+	"github.com/redhat-developer/odo/pkg/preference"
+	"github.com/redhat-developer/odo/pkg/segment"
+	"github.com/redhat-developer/odo/pkg/testingutil/filesystem"
+	"github.com/redhat-developer/odo/pkg/util"
+
+	"k8s.io/kubectl/pkg/util/templates"
+)
+
+// RecommendedCommandName is the recommended command name
+const RecommendedCommandName = "init"
+
+var initExample = templates.Examples(`
+  # Boostrap a new component in interactive mode
+  %[1]s
+`)
+
+type InitOptions struct {
+	// Backends to build init parameters
+	backends []params.ParamsBuilder
+
+	// filesystem on which command is running
+	fsys filesystem.Filesystem
+
+	// Clients
+	preferenceClient preference.Client
+	registryClient   registry.Client
+
+	// the parameters needed to run the init procedure
+	params.InitParams
+
+	// Destination directory
+	destDir string
+}
+
+// NewInitOptions creates a new InitOptions instance
+func NewInitOptions(backends []params.ParamsBuilder, fsys filesystem.Filesystem, prefClient preference.Client, registryClient registry.Client) *InitOptions {
+	return &InitOptions{
+		backends:         backends,
+		fsys:             fsys,
+		preferenceClient: prefClient,
+		registryClient:   registryClient,
+		destDir:          ".",
+	}
+}
+
+// Complete will build the parameters for init, using different backends based on the flags set,
+// either by using flags or interactively if no flag is passed
+// Complete will return an error immediately if the current working directory is not empty
+func (o *InitOptions) Complete(cmdline cmdline.Cmdline, args []string) (err error) {
+
+	empty, err := isEmpty(o.fsys, o.destDir)
+	if err != nil {
+		return err
+	}
+	if !empty {
+		return errors.New("The current directory is not empty. You can bootstrap new component only in empty directory.\nIf you have existing code that you want to deploy use `odo deploy` or use `odo dev` command to quickly iterate on your component.")
```

Comment:
Yes, that could be nice. I would prefer we decide on a way to do this all around the code before to start doing it in some packages

Reply:
I suggested we do it here since we are writing things relatively from scratch, and it would be easier to add these changes as well, but sure we can discuss a decisive way of doing this.